[Week5] 고양이 장난감가게 만들기

yongjunleeme opened the pull request:
안녕하세요! 잘 부탁을 드립니다~

yongjunleeme committed:
Add User, UserRepository, JpaUserRepository

yongjunleeme committed:
Add getUserTest in UserServiceTest

yongjunleeme committed:
Add createUser in UserService

yongjunleeme committed:
Update build.gradle

yongjunleeme committed:
Add updateUser in UserService

yongjunleeme commented:
`updateUser` 테스트 중에 `getUser()`를 호출할 때 `UserNotFoundException` 예외를 극복하지 못하고 PR을 드렸습니다;;

어떻게 해야 할지 힌트라도 여쭈어 볼 수 있을까요~?

ssisksl77 commented:
> `updateUser` 테스트 중에 `getUser()`를 호출할 때 `UserNotFoundException` 예외를 극복하지 못하고 PR을 드렸습니다;;
> 
> 어떻게 해야 할지 힌트라도 여쭈어 볼 수 있을까요~?

어이쿠 당연히 도와드려야죠! 
코드 코멘트에 남겨놓을게요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,201 @@
+
+            @Test
+            @DisplayName("업데이트된 사용자를 리턴한다")
+            void it_returns_updated_user() {
+                User updateUser = userService.updateUser(STORED_ID, updateUserData);
```
```suggestion
            void it_returns_updated_user() {
                given(userRepository.findById(STORED_ID)).willReturn(Optional.of(user));
                User updateUser = userService.updateUser(STORED_ID, updateUserData);
```

UserRepository를 모킹하셨었네요! userRepository가 원하는 방식으로 작동하도록 해줘야해요!

yongjunleeme replied:
헉!! userRepository가 원하는 방식이 무엇인지 구체적으로 조금만 더 힌트를 주시면 안 될까요.. 여기서 24시간 정체 중입니다..!

yongjunleeme replied:
헉헉!! 해결하였습니다;;

ssisksl77 replied:
어이쿠 제가 이제야 봤군요 ㅠㅜㅠㅜ 고생하셨습니다.

ssisksl77 replied:
제가 제안한방식으로 복붙하면 되지않았을까 싶은데 잘 안 되셨나보군요 ㅠㅜㅠㅜ

yongjunleeme replied:
헉헉헉!! 아닙니다. 고이 써주신 코드를 보지도 못하고 `userRepository가 원하는 방식` 코멘트에 대해서만 한참 고민하다가 삽질을 하였습니다..ㅎㅎ; 

알려주셔서 감사드립니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/infra/JpaUserRepository.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+import java.util.Optional;
+
+@Primary
```
`@Primary` 가 없으면 어떻게 될까요?!

yongjunleeme replied:
2개를 쓸 때 우선권을 주는 용도로 알고 있습니다! 없어도 될 것 같습니다 ㅎㅎ

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/dto/UserData.java:
```diff
@@ -0,0 +1,34 @@
+    private String email;
+
+    @Mapping("password")
+    @NotBlank(message = "password 값은 필수입니다")
```
좋아요! 한번 커스텀 Validator를 만들어서 `@NotBlank` 를 대체해보아요!

yongjunleeme replied:
헉, 넵 위에 모킹 에러를 잡고 도전해야 할 것 같습니다~

yongjunleeme replied:
Custom Validator를 [시도](https://github.com/CodeSoom/spring-week5-assignment-1/pull/67/commits/e5c89e7c85941a77c396c21c1d5366deab4e36bf)해 보았는데 작동이 안 됩니다 ㅠㅠ~
힌트를 부탁드릴 수 있을까요~?

참조한 블로그
- https://ugo04.tistory.com/87
- https://www.baeldung.com/spring-mvc-custom-validator#custom-validation

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,81 @@
+     * @param userData - 수정할 사용자
+     * @return 수정한 사용자
+     */
+    public User updateUser(Long id, UserData userData) {
```
우리는 방금 updateUser가 `UserNotFoundException` 을 던지는 것을 확인했어요!

주석에 설명이 필요할 것 같아요!

yongjunleeme replied:
넵넵 알겠습니다~ `findUser()` 메서드 문서에 `UserNotFoundException` 을 명시하더라도 `findUser()`를 가져다 쓰는 다른 메서드에도 같은 에러 설명을 덧붙여줘야 하나요~?

ssisksl77 replied:
맞아요 좋은 의문이에요! 왜냐하면 이것에 정답은 없기 때문입니다!
문서는 상대가 잘 이해하길 또 미래에 내가 이 코드를 잘 이해하게 하기위한 마음이라고 생각합니다. 

그럼에도 한번 진지하게 생각해봅시다! 

만약에 `findUser` 로 메소드를 분리하지 않고 inline으로 사용하면 해당 예외는 문서에 추가해야 하는 걸까?
리팩토링을 하면서 당연히 문서도 분리되는 것이 아닌가? 
라는 의문이 생깁니다.

하지만 문서가 분리된다는 것은 서로가 하는 일이 영향을 주지 않을 때일 것입니다.

현재 `updateUser` 는 `UserNotFoundException` 를 던지는 `findUser` 를 사용하는 것을 앎에도 예외를 핸들링 하지 않았습니다. 
(이 말을 이 예외는 내가 회복시킬 수 없다는 의미라고 해봅시다)

그렇다면 이 책은 다시 회복시키지 않은 `updateUser` 가 예외를 던지는 것이죠!!

`updateUser` 는 예외를 만났을 때 3가지 옵션이 있었습니다.
 1. 새로운 예외를 만들어서 던진다.
 2. try/catch 로 회복시킨다
 3. 그대로 다시 던진다

1, 2 번으로 했다면 분명 문서에 적혀있었을 것입니다. 그리고 3번이라고 예외는 없을 것입니다.

yongjunleeme replied:
오오 가르침 감사드립니다!

정답은 없지만 메소드 사용자 관점에서 메소드 책임에 해당하는 예외는 해당 메소드의 문서에 정의해주는 것이 좋다. `UserNotFoundException` 책임이 `updateUser`에게 있으니 `updateUser`가 에러를 핸들링할 수 없더라도 3번 옵션에 의해 그대로 다시 던지는 내용을 사용자에게 알려주어야 한다.

라고 이해했습니다~!

yongjunleeme committed:
Fix updateUserTest in UserServiceTest

yongjunleeme committed:
Add deleteUser in UserServiceTest

yongjunleeme committed:
Add userTest

yongjunleeme committed:
Add password custom Validator

yongjunleeme committed:
Add findUser in UserControllerTest

yongjunleeme committed:
Fix findUser in UserControllerTest

yongjunleeme commented:
UserControllerTest에서 mockMvc로 findUser를 하는데 null 응답을 받고 있습니다;; UserServiceTest에서 처럼 제가 Mock 테스트 과정을 잘 몰라 계속 삽질을 하는 것 같아서.. 도움을 요청드립니다!!

yongjunleeme commented:
1)
추가로 혹시 데이터가 DB에 저장되는 시점은 어떤 코드에서 격발되는 것인지 여쭈어볼 수 있을까요~?

`UserRepository`의 `createUser()` 메소드를 호출하면`userRepository.save(user)`에서 메모리가 아닌 영속화된 DB에 저장한다고 생각했는데

그럼 `updateUser`에서 save를 호출하지 않으면 저장이 안 되는 것인가 싶었습니다~

2)
그리고 Spring DB를 공부하려면 SQL문법과 JPA 중 무엇이 더 중요한지.. 여쭤봐도 될까요~? 저는 그전에 Django를 썼는데 ORM 문법 몇 개만 알면 불편하지 않은 수준이었습니다. 하지만 아샬님의 [영상](https://www.youtube.com/watch?v=ObelJbEt8-Y&ab_channel=%EC%BD%94%EB%94%A9%EC%9D%98%EC%8B%A0%EC%95%84%EC%83%AC)을 보고 SQL을 쓰지 않는 단점을 극복하고자 [RealMysql](http://www.yes24.com/Product/Goods/103415767) 책을 보고 SQL 쿼리를 정리해왔습니다. 그런데 막상 쓸 기회가 없으니 기억에서 휘발되곤 했습니다.

자바 스프링으로 취업하면 SQL을 쓸 줄 알았는데 JPA를 알아야 한다는 것 같아서 제가 또 애초에 잘못된 추론으로 접근해서 시간을 낭비할까 싶어 질문을 드려봅니다~!

ssisksl77 commented:
1. 당연히 값은 변경했으면 `save` 를 해야합니다! 하지만 우리에겐 그것마저 귀찮지요! 그렇기 때문에 Spring Data에서는 dirty checking이 대신하도록 할 수 있습니다! 이 Spring Data dirty checking에 대해서 알아보고 이곳에 공유해주세요! (출처와 함께요!)
2. 너무 식상한 답변이지만 둘다 중요합니다. ReadMySql을 읽어보셨다니 현재는 JPA에 집중하셔도 좋을 것 같아요! (JPA는 SQL을 대신 작성해줍니다. 그러므로 점점 시간이 지남에 따라 SQL이 필요해질 거에요!)

ssisksl77 commented:
> UserControllerTest에서 mockMvc로 findUser를 하는데 null 응답을 받고 있습니다;; UserServiceTest에서 처럼 제가 Mock 테스트 과정을 잘 몰라 계속 삽질을 하는 것 같아서.. 도움을 요청드립니다!!

안녕하세요! 처음에는 문제를 해결하도록 가이드를 드릴까 했는데요! 
이 문제는 가이드를 드리는 것보다 빠르게 해결하여 개발하면서 더 좋은 경험을 드리는 것이 좋겠다고 판단하였습니다!

대신 해결한 곳을 보시고 해당 코드에 대해서 좀 더 자세히 알아보는 시간을 갖고 이곳에 댓글로 공유를 해주세요! (알아본출처도 함께요!)

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,81 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@DisplayName("UserController 클래스")
+@WebMvcTest(UserControllerTest.class)
```
```suggestion
@WebMvcTest(UserController.class)
```
이렇게 변경하시면 해결될거에요!

댓글에 써있듯이 이 코드에 어노테이션에 대해서 심도있게 찾아보시고 공유해주세요!

yongjunleeme replied:
넵 잘 알겠습니다. 도움주셔서 감사드립니다!

## `@WebMvcTest()`
 
- `@Controller`, `@ControllerAdvice` 등의 빈을 스캔한다. [자동 설정 목록](https://docs.spring.io/spring-boot/docs/current/reference/html/test-auto-configuration.html#appendix.test-auto-configuration)  
	- `@WebMvcTest(UserController.class)`: 괄호 안에 테스트할 클래스 명시할 수도 있음
- `MockMvc`로 컨트롤러를 자동 설정에 추가한다.
	- 하나의 컨트롤러로 제한하지 않으려면 `@MockBean`으로 등록한다. 

```java

@WebMvcTest(UserController.class)  
class UserControllerTest {

@Autowired  
private MockMvc mockMvc;
  
@MockBean  
private UserService userService;
}
```
 
### 장점
- 웹 애플리케이션 관련 빈만 등록하므로 빠르다.
-  통합 테스트를 진행하기 어려운 테스트를 진행 가능하다.
ex) 결제 모듈 API를 콜하며 안되는 상황에서 Mock을 통해 가짜 객체를 만들어 테스트 가능.

### 단점

- 요청부터 응답까지 모든 테스트를 Mock 기반으로 테스트하기 때문에 실제 환경에서는 제대로 동작하지 않을 수 있다.

### 예제

```java
@Test  
@DisplayName("찾은 사용자와 상태코드 200을 응답한다")  
void it_responds_found_user_and_status_code_200() throws Exception {  
    //given  
    given(userService.getUser(EXISTING_ID)).willReturn(user);  
    //when  
    mockMvc.perform(get("/users/{id}", EXISTING_ID)  
                    .accept(MediaType.APPLICATION_JSON)  
                    .contentType(MediaType.APPLICATION_JSON))  
              
            // then  
            .andExpect(jsonPath("$.id").exists())  
            .andExpect(jsonPath("name").exists())  
            .andExpect(jsonPath("email").exists())  
            .andExpect(status().isOk());  
}
```

- given: Mock 객체 생성
- when: Mock 객체 반환
- then: 응답 검사

## 출처
- [sprint boot doc](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.testing.spring-boot-applications.spring-mvc-tests)
- [갓대희의 작은공간:티스토리](https://goddaehee.tistory.com/212)

ssisksl77 replied:
좋아요!

yongjunleeme commented:
## dirty checking

- JPA의 상태변화 감지 자동 저장 기능?
- JPA에서 엔티티 조회 시 스냅샷 생성
- 영속성 컨텍스트 내에서 스냅샷과 다른 상태 변화 감지 시 업데이트 쿼리 호출
- 전체 필드를 업데이트 하는 방식
- 개별 필드를 업데이트하려면 `@DynamicUpdate`를 사용

참고: https://jojoldu.tistory.com/415

공식문서를 참조하고 싶었는데 [jpa 문서](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/)에서 dirty checking을 찾을 수 없었습니다. (공식 문서가 아닌가요..??) [Hibernate 문서](https://docs.jboss.org/hibernate/orm/6.0/userguide/html_single/Hibernate_User_Guide.html)에는 dirty checking이 언급되는데 과제 코드에서는 Hibernate를 안 쓴다고 생각했습니다. 이에 검색해서 위 참고 블로그를 보니 내용의 원본은 저번주와 같이 [자바 ORM 표준 JPA 프로그래밍](https://www.aladin.co.kr/shop/wproduct.aspx?ItemId=62681446) 책을 참조했다고 합니다. 모든 길은 이 책으로 통하는 것인지...ㅎㅎ;

과제 코드에서 dirty checking 로직을 어디서 호출하나 찾아보고 싶었습니다. `JpaUserRepository` 인터페이스에서 `CrudRepository`를 상속하여 SpringData에 접근한다고 생각했는데 막상 `CrudRepository`에 들어가보니 인터페이스만 있고 실행 코드는 찾을 수 없었습니다.
이 [블로그](https://blog.woniper.net/346?category=699184%20%EC%B6%9C%EC%B2%98:%20https://xzio.tistory.com/646%20[%EC%BD%94%EB%94%A9%EC%B0%BD%EA%B3%A0:%ED%8B%B0%EC%8A%A4%ED%86%A0%EB%A6%AC) 를 참조하니 CrudRepository가 [SimpleJpaRepository](http://docs.spring.io/autorepo/docs/spring-data-jpa/1.6.2.RELEASE/api/org/springframework/data/jpa/repository/support/SimpleJpaRepository.html) 구현체에 의해 동작한다는데 CrudRepository에서 어떻게 이어져 있는지는 잘 모르겠습니다. 일단 SimpleJpaRepository 코드를 보니 save 비슷한 코드가 있어서 여기서 작동을 하는 것인가 싶었습니다~!

<img width="678" alt="simpleJpaRepository" src="https://user-images.githubusercontent.com/48748376/171967039-0f364cac-d3c2-470a-b161-93cd5e268c70.png">

## 질문

- CrudRepository가 [SimpleJpaRepository](http://docs.spring.io/autorepo/docs/spring-data-jpa/1.6.2.RELEASE/api/org/springframework/data/jpa/repository/support/SimpleJpaRepository.html) 구현체에 의해 동작한다는데 CrudRepository에서 어떻게 이어지는지 여쭤볼 수 있을까요~?

yongjunleeme commented:
Custom Validator를 [시도](https://github.com/CodeSoom/spring-week5-assignment-1/pull/67/commits/e5c89e7c85941a77c396c21c1d5366deab4e36bf)해 보았는데 작동이 안 됩니다 ㅠㅠ~
힌트를 부탁드릴 수 있을까요~?

참조한 블로그

https://ugo04.tistory.com/87
https://www.baeldung.com/spring-mvc-custom-validator#custom-validation

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+}
```
EOF에 빈칸을 추가해주세요!

ssisksl77 commented:
> Custom Validator를 [시도](https://github.com/CodeSoom/spring-week5-assignment-1/pull/67/commits/e5c89e7c85941a77c396c21c1d5366deab4e36bf)해 보았는데 작동이 안 됩니다 ㅠㅠ~ 힌트를 부탁드릴 수 있을까요~?
> 
> 참조한 블로그
> 
> https://ugo04.tistory.com/87 https://www.baeldung.com/spring-mvc-custom-validator#custom-validation

작동이 안된다는 것보다는 어떤 상황에서 어떤 것이

> Custom Validator를 [시도](https://github.com/CodeSoom/spring-week5-assignment-1/pull/67/commits/e5c89e7c85941a77c396c21c1d5366deab4e36bf)해 보았는데 작동이 안 됩니다 ㅠㅠ~ 힌트를 부탁드릴 수 있을까요~?
> 
> 참조한 블로그
> 
> https://ugo04.tistory.com/87 https://www.baeldung.com/spring-mvc-custom-validator#custom-validation

어떤 환경에 무엇을 호출했더니 예상했던 결과가 A였지만 B가 나왔다는식의 설명이 있으면 도와드리는데 큰 도움이 될 것 같아요!

실패하는 테스트를 작성해서 올려주세요!

ssisksl77 commented:
> ## dirty checking
> * JPA의 상태변화 감지 자동 저장 기능?
> * JPA에서 엔티티 조회 시 스냅샷 생성
> * 영속성 컨텍스트 내에서 스냅샷과 다른 상태 변화 감지 시 업데이트 쿼리 호출
> * 전체 필드를 업데이트 하는 방식
> * 개별 필드를 업데이트하려면 `@DynamicUpdate`를 사용
> 
> 참고: https://jojoldu.tistory.com/415
> 
> 공식문서를 참조하고 싶었는데 [jpa 문서](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/)에서 dirty checking을 찾을 수 없었습니다. (공식 문서가 아닌가요..??) [Hibernate 문서](https://docs.jboss.org/hibernate/orm/6.0/userguide/html_single/Hibernate_User_Guide.html)에는 dirty checking이 언급되는데 과제 코드에서는 Hibernate를 안 쓴다고 생각했습니다. 이에 검색해서 위 참고 블로그를 보니 내용의 원본은 저번주와 같이 [자바 ORM 표준 JPA 프로그래밍](https://www.aladin.co.kr/shop/wproduct.aspx?ItemId=62681446) 책을 참조했다고 합니다. 모든 길은 이 책으로 통하는 것인지...ㅎㅎ;
> 
> 과제 코드에서 dirty checking 로직을 어디서 호출하나 찾아보고 싶었습니다. `JpaUserRepository` 인터페이스에서 `CrudRepository`를 상속하여 SpringData에 접근한다고 생각했는데 막상 `CrudRepository`에 들어가보니 인터페이스만 있고 실행 코드는 찾을 수 없었습니다. 이 [블로그](https://blog.woniper.net/346?category=699184%20%EC%B6%9C%EC%B2%98:%20https://xzio.tistory.com/646%20%5B%EC%BD%94%EB%94%A9%EC%B0%BD%EA%B3%A0:%ED%8B%B0%EC%8A%A4%ED%86%A0%EB%A6%AC) 를 참조하니 CrudRepository가 [SimpleJpaRepository](http://docs.spring.io/autorepo/docs/spring-data-jpa/1.6.2.RELEASE/api/org/springframework/data/jpa/repository/support/SimpleJpaRepository.html) 구현체에 의해 동작한다는데 CrudRepository에서 어떻게 이어져 있는지는 잘 모르겠습니다. 일단 SimpleJpaRepository 코드를 보니 save 비슷한 코드가 있어서 여기서 작동을 하는 것인가 싶었습니다~!
> 
> <img alt="simpleJpaRepository" width="678" src="https://user-images.githubusercontent.com/48748376/171967039-0f364cac-d3c2-470a-b161-93cd5e268c70.png">
> 
> ## 질문
> * CrudRepository가 [SimpleJpaRepository](http://docs.spring.io/autorepo/docs/spring-data-jpa/1.6.2.RELEASE/api/org/springframework/data/jpa/repository/support/SimpleJpaRepository.html) 구현체에 의해 동작한다는데 CrudRepository에서 어떻게 이어지는지 여쭤볼 수 있을까요~?

- `SimpleJpaRepository` 는 CrudRepository의 기본 구현체입니다! 다른 설정이 추가되지 않는다면 `SimpleJpaRepository` 를 사용할 것입니다!

ssisksl77 commented:
코드를 받아서 실행해보았는데 App 실행이 안되네요! 

이것부터 해결을 해야겠어요!

yongjunleeme commented:
> 코드를 받아서 실행해보았는데 App 실행이 안되네요!
> 
> 이것부터 해결을 해야겠어요!

헉 넵넵 그런데 시간이.. 덜덜 ㅜㅜ 이번주에는 이례적으로 강의 풀이 영상이 무려 2.5시간이고 이걸 보는 것뿐만 아니라 회상해서 구현도 해야 하고 회고도 쓰고 제 과제도 완성을 해야 하는데 시간이 부족할 것 같습니다.. (할 수 있는 데까지 해보겠습니다~!)

yongjunleeme committed:
Fix UserNotFouncException

yongjunleeme committed:
Fix User

yongjunleeme committed:
Update UseControllerTest

yongjunleeme commented:
> ## dirty checking
> * JPA의 상태변화 감지 자동 저장 기능?
> * JPA에서 엔티티 조회 시 스냅샷 생성
> * 영속성 컨텍스트 내에서 스냅샷과 다른 상태 변화 감지 시 업데이트 쿼리 호출
> * 전체 필드를 업데이트 하는 방식
> * 개별 필드를 업데이트하려면 `@DynamicUpdate`를 사용
> 
> 참고: https://jojoldu.tistory.com/415
> 
> 공식문서를 참조하고 싶었는데 [jpa 문서](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/)에서 dirty checking을 찾을 수 없었습니다. (공식 문서가 아닌가요..??) [Hibernate 문서](https://docs.jboss.org/hibernate/orm/6.0/userguide/html_single/Hibernate_User_Guide.html)에는 dirty checking이 언급되는데 과제 코드에서는 Hibernate를 안 쓴다고 생각했습니다. 이에 검색해서 위 참고 블로그를 보니 내용의 원본은 저번주와 같이 [자바 ORM 표준 JPA 프로그래밍](https://www.aladin.co.kr/shop/wproduct.aspx?ItemId=62681446) 책을 참조했다고 합니다. 모든 길은 이 책으로 통하는 것인지...ㅎㅎ;
> 
> 과제 코드에서 dirty checking 로직을 어디서 호출하나 찾아보고 싶었습니다. `JpaUserRepository` 인터페이스에서 `CrudRepository`를 상속하여 SpringData에 접근한다고 생각했는데 막상 `CrudRepository`에 들어가보니 인터페이스만 있고 실행 코드는 찾을 수 없었습니다. 이 [블로그](https://blog.woniper.net/346?category=699184%20%EC%B6%9C%EC%B2%98:%20https://xzio.tistory.com/646%20%5B%EC%BD%94%EB%94%A9%EC%B0%BD%EA%B3%A0:%ED%8B%B0%EC%8A%A4%ED%86%A0%EB%A6%AC) 를 참조하니 CrudRepository가 [SimpleJpaRepository](http://docs.spring.io/autorepo/docs/spring-data-jpa/1.6.2.RELEASE/api/org/springframework/data/jpa/repository/support/SimpleJpaRepository.html) 구현체에 의해 동작한다는데 CrudRepository에서 어떻게 이어져 있는지는 잘 모르겠습니다. 일단 SimpleJpaRepository 코드를 보니 save 비슷한 코드가 있어서 여기서 작동을 하는 것인가 싶었습니다~!
> 
> <img alt="simpleJpaRepository" width="678" src="https://user-images.githubusercontent.com/48748376/171967039-0f364cac-d3c2-470a-b161-93cd5e268c70.png">

`User`의 password를 영문, 특수문자, 숫자 포함 8자리 이상으로 유효성 체크를 하기 위해서 `PasswordValidator', `Password` 를 생성하였고 `UserTest`에서 password를 1234로만 넣어서 빌드를 해보았는데 작동을 하지 않아서 질문을 남겨 놓았습니다!!;;

yongjunleeme commented:
> > Custom Validator를 [시도](https://github.com/CodeSoom/spring-week5-assignment-1/pull/67/commits/e5c89e7c85941a77c396c21c1d5366deab4e36bf)해 보았는데 작동이 안 됩니다 ㅠㅠ~ 힌트를 부탁드릴 수 있을까요~?
> > 참조한 블로그
> > https://ugo04.tistory.com/87 https://www.baeldung.com/spring-mvc-custom-validator#custom-validation
> 
> 작동이 안된다는 것보다는 어떤 상황에서 어떤 것이
> 
> > Custom Validator를 [시도](https://github.com/CodeSoom/spring-week5-assignment-1/pull/67/commits/e5c89e7c85941a77c396c21c1d5366deab4e36bf)해 보았는데 작동이 안 됩니다 ㅠㅠ~ 힌트를 부탁드릴 수 있을까요~?
> > 참조한 블로그
> > https://ugo04.tistory.com/87 https://www.baeldung.com/spring-mvc-custom-validator#custom-validation
> 
> 어떤 환경에 무엇을 호출했더니 예상했던 결과가 A였지만 B가 나왔다는식의 설명이 있으면 도와드리는데 큰 도움이 될 것 같아요!
> 
> 실패하는 테스트를 작성해서 올려주세요!

User의 password를 영문, 특수문자, 숫자 포함 8자리 이상으로 유효성 체크를 하기 위해서 PasswordValidator', Password를 생성하였고UserTest`에서 password를 1234로만 넣어서 빌드를 해보았는데 작동을 하지 않아서 질문을 남겨 놓았습니다!!;;

yongjunleeme commented:
UserControllerTest의 84번 라인부터 저장되지 않은 사용자 id가 주어질 때 테스트를 하는데 해결이 안 되어서 또 질문 의존을 드립니다 ㅠ.ㅠ~;;

given이 Mock 객체 생성을 하는 수순이고 아래와 같이 하는 게 맞지 않을까요..? 

```java
@BeforeEach
            void setUp() {
                given(userService.getUser(NOT_EXISTING_ID))
                        .willThrow(new UserNotFoundException());
            }
```

ssisksl77 commented:
> UserControllerTest의 84번 라인부터 저장되지 않은 사용자 id가 주어질 때 테스트를 하는데 해결이 안 되어서 또 질문 의존을 드립니다 ㅠ.ㅠ~;;
> 
> given이 Mock 객체 생성을 하는 수순이고 아래와 같이 하는 게 맞지 않을까요..?
> 
> ```java
> @BeforeEach
>             void setUp() {
>                 given(userService.getUser(NOT_EXISTING_ID))
>                         .willThrow(new UserNotFoundException());
>             }
> ```

잘하셨는데요!!! ㅎㅎㅎ 
리턴값도 `UserNotFoundException()` 으로 잘 던져지고 있네요!
문제는 왜! 404로 리턴하지 않느냐는 거겠죠?

그것은 아직 UserNotFoundException을 404로 변환하지 않았기 때문이에요!

ssisksl77 commented:
> PasswordValidator

아~ UserTest 내에서 PASSWORD 가 1234 임에도 왜 인스턴스가 정상적으로 만들어졌냐는 말씀이시군요!
그것은 유효성검증을 하지 않았기 때문입니다. 스프링은 `@Valid` 어노테이션을 보고 선별적으로 유효성검증을 하지만 이 코드는 그냥 인스턴스를 만드는 것이므로 유효성검증을 하지 않았습니다.

유효성검증 자체 java Bean Validation은 Spring과 관련없으므로 직접호출해서 사용하는 법도 찾아보면 좋을 것 같아요!

yongjunleeme commented:
> > UserControllerTest의 84번 라인부터 저장되지 않은 사용자 id가 주어질 때 테스트를 하는데 해결이 안 되어서 또 질문 의존을 드립니다 ㅠ.ㅠ~;;
> > given이 Mock 객체 생성을 하는 수순이고 아래와 같이 하는 게 맞지 않을까요..?
> > ```java
> > @BeforeEach
> >             void setUp() {
> >                 given(userService.getUser(NOT_EXISTING_ID))
> >                         .willThrow(new UserNotFoundException());
> >             }
> > ```
> 
> 잘하셨는데요!!! ㅎㅎㅎ 리턴값도 `UserNotFoundException()` 으로 잘 던져지고 있네요! 문제는 왜! 404로 리턴하지 않느냐는 거겠죠?
> 
> 그것은 아직 UserNotFoundException을 404로 변환하지 않았기 때문이에요!

아하 그렇군요! 넵넵 감사드립니다~!

yongjunleeme commented:
> > PasswordValidator
> 
> 아~ UserTest 내에서 PASSWORD 가 1234 임에도 왜 인스턴스가 정상적으로 만들어졌냐는 말씀이시군요! 그것은 유효성검증을 하지 않았기 때문입니다. 스프링은 `@Valid` 어노테이션을 보고 선별적으로 유효성검증을 하지만 이 코드는 그냥 인스턴스를 만드는 것이므로 유효성검증을 하지 않았습니다.
> 
> 유효성검증 자체 java Bean Validation은 Spring과 관련없으므로 직접호출해서 사용하는 법도 찾아보면 좋을 것 같아요!

아하 이 타래는 트레이너님께서 말씀주신 custom validation을 적용하고자 시도했던 것인데.. 해결이 안 되네요 ㅠㅠ 나중에 방법을 알게되면 적용해보겠습니다!! 감사합니다~!

yongjunleeme committed:
미해결 오류 일단 삭제

- password에 custom validation 적용
- UserControllerTest에서 받는 오류
    - `org.springframework.web.util.NestedServletException: Request
      processing failed; nested exception is
      com.codesoom.assignment.UserNotFoundException: 사용자를 찾을 수
      없습니다.`

yongjunleeme commented:
[5주차 회고](https://github.com/yongjunleeme/spring-code-soom/blob/main/docs/week5.MD)입니다~

ssisksl77 commented:
> > > PasswordValidator
> > 
> > 
> > 아~ UserTest 내에서 PASSWORD 가 1234 임에도 왜 인스턴스가 정상적으로 만들어졌냐는 말씀이시군요! 그것은 유효성검증을 하지 않았기 때문입니다. 스프링은 `@Valid` 어노테이션을 보고 선별적으로 유효성검증을 하지만 이 코드는 그냥 인스턴스를 만드는 것이므로 유효성검증을 하지 않았습니다.
> > 유효성검증 자체 java Bean Validation은 Spring과 관련없으므로 직접호출해서 사용하는 법도 찾아보면 좋을 것 같아요!
> 
> 아하 이 타래는 트레이너님께서 말씀주신 custom validation을 적용하고자 시도했던 것인데.. 해결이 안 되네요 ㅠㅠ 나중에 방법을 알게되면 적용해보겠습니다!! 감사합니다~!

아하 직접 객체의 유효성검증을 호출해야 한다는 말이었습니다!
아마도 해당 유효성검증을 테스트하려면 두가지가 있겠죠! 
- 통합테스트나, WebMvcTest로 유효성검증 테스트를 만든다.
- 직접 Validator를 호출하여 객체의 유효성검증을 수행한다!

yongjunleeme commented:
> `@Valid`

직접 Validator를 호출하여 유효성검증을 수행한다!
이 방법으로 시도했던 것인데 `@Valid`를 사용해 스프링을 쓰는 게 아니라 유효성검증 자체 java Bean Validation은 Spring과 관련없으므로 직접호출해서 사용하는 법을 제가 알아봐야 했던 것이군요~! 잘 알겠습니다 감사합니다!

ssisksl77 commented:
> > `@Valid`
> 
> 직접 Validator를 호출하여 유효성검증을 수행한다! 이 방법으로 시도했던 것인데 `@Valid`를 사용해 스프링을 쓰는 게 아니라 유효성검증 자체 java Bean Validation은 Spring과 관련없으므로 직접호출해서 사용하는 법을 제가 알아봐야 했던 것이군요~! 잘 알겠습니다 감사합니다!

조금 도움을 드리자면 이 링크가 힌트가 될 수 있겠습니다!
https://docs.oracle.com/javaee/7/api/javax/validation/Validation.html

yongjunleeme commented:
> > > `@Valid`
> > 
> > 
> > 직접 Validator를 호출하여 유효성검증을 수행한다! 이 방법으로 시도했던 것인데 `@Valid`를 사용해 스프링을 쓰는 게 아니라 유효성검증 자체 java Bean Validation은 Spring과 관련없으므로 직접호출해서 사용하는 법을 제가 알아봐야 했던 것이군요~! 잘 알겠습니다 감사합니다!
> 
> 조금 도움을 드리자면 이 링크가 힌트가 될 수 있겠습니다! https://docs.oracle.com/javaee/7/api/javax/validation/Validation.html

아니 이런 귀한 링크를! 감사합니다!!
